File: scratch/gen_theoretical.js
const fs = require('fs');
const path = require('path');

// Ported from theoreticalSegmentSelector.ts and theoreticalLapAssembler.ts
function selectBestSegments(laps, segments) {
  const bestSectors = [];
  const segmentGroups = {};

  segments.forEach(seg => {
    const baseId = seg.segment_id.split('_')[0]; 
    if (!segmentGroups[baseId]) segmentGroups[baseId] = [];
    segmentGroups[baseId].push(seg);
  });

  for (const baseId in segmentGroups) {
    const sorted = segmentGroups[baseId].sort((a, b) => a.duration_seconds - b.duration_seconds);
    bestSectors.push(sorted[0]);
  }

  return bestSectors.sort((a, b) => a.start_timestamp - b.start_timestamp);
}

function assembleTheoreticalLap(telemetry, bestSectors) {
  const resultTelemetry = [];
  let currentTimestamp = 0;

  bestSectors.forEach(seg => {
    const slice = telemetry.filter(t => t.timestamp >= seg.start_timestamp && t.timestamp <= seg.end_timestamp);
    const sliceStart = slice[0].timestamp;

    slice.forEach(t => {
      resultTelemetry.push({
        ...t,
        original_timestamp: t.timestamp,
        timestamp: currentTimestamp + (t.timestamp - sliceStart),
      });
    });

    currentTimestamp += (seg.end_timestamp - seg.start_timestamp);
  });

  return {
    telemetry: resultTelemetry,
    totalDurationMs: currentTimestamp,
    sectors: bestSectors
  };
}

async function run() {
  const telemetry = JSON.parse(fs.readFileSync('frontend/public/demo/telemetry.json', 'utf8'));
  const laps = JSON.parse(fs.readFileSync('frontend/public/demo/laps.json', 'utf8'));
  const segments = JSON.parse(fs.readFileSync('frontend/public/demo/segments.json', 'utf8'));

  const bestSectors = selectBestSegments(laps, segments);
  const tLap = assembleTheoreticalLap(telemetry, bestSectors);

  fs.writeFileSync('frontend/public/demo/theoretical_best.json', JSON.stringify(tLap, null, 2));
  console.log('✅ Generated theoretical_best.json');
}

run();

File: scratch/gen_theoretical.py
import json
import os

def select_best_segments(laps, segments):
    segment_groups = {}
    for seg in segments:
        base_id = seg['segment_id'].split('_')[0]
        if base_id not in segment_groups:
            segment_groups[base_id] = []
        segment_groups[base_id].append(seg)
    
    best_sectors = []
    for base_id in segment_groups:
        sorted_segs = sorted(segment_groups[base_id], key=lambda x: x['duration_seconds'])
        best_sectors.append(sorted_segs[0])
    
    return sorted(best_sectors, key=lambda x: x['start_timestamp'])

def assemble_theoretical_lap(telemetry, best_sectors):
    result_telemetry = []
    current_timestamp = 0
    
    for seg in best_sectors:
        slice_points = [t for t in telemetry if seg['start_timestamp'] <= t['timestamp'] <= seg['end_timestamp']]
        if not slice_points:
            continue
            
        slice_start = slice_points[0]['timestamp']
        for t in slice_points:
            new_pt = t.copy()
            new_pt['original_timestamp'] = t['timestamp']
            new_pt['timestamp'] = current_timestamp + (t['timestamp'] - slice_start)
            result_telemetry.append(new_pt)
            
        current_timestamp += (seg['end_timestamp'] - seg['start_timestamp'])
        
    return {
        "telemetry": result_telemetry,
        "totalDurationMs": current_timestamp,
        "sectors": best_sectors
    }

def run():
    with open('frontend/public/demo/telemetry.json', 'r') as f:
        telemetry = json.load(f)
    with open('frontend/public/demo/laps.json', 'r') as f:
        laps = json.load(f)
    with open('frontend/public/demo/segments.json', 'r') as f:
        segments = json.load(f)
        
    best_sectors = select_best_segments(laps, segments)
    t_lap = assemble_theoretical_lap(telemetry, best_sectors)
    
    with open('frontend/public/demo/theoretical_best.json', 'w') as f:
        json.dump(t_lap, f, indent=2)
    print("✅ Generated theoretical_best.json")

if __name__ == "__main__":
    run()
